### DISCUSSION TITLE
zfs import incomplete at boot

### DISCUSSION THREAD

**Discussion by bazald** [openzfs/zfs#8960] [Q&A] [answered]
### System information
Type	|	Version/Name
---	|	---
Distribution Name	|	Debian
Distribution Version	|	Stretch 
Linux Kernel	|	4.19.55 #1 SMP PREEMPT
Architecture	|	x86_64
ZFS Version	|	0.8.1-1
SPL Version	|	0.8.1-1

### Describe the problem you're observing
At boot, zfs import is clearly being attempted, because zfs import reports that the pool has already been imported. However, the pool is inaccessible until I do the following dance:

1. zfs export pool
2. rmdir /pool (the mountpoint remains present for some reason and blocks import)
3. zfs import pool

After that, the pool is accessible and services depending on the pool can be started successfully.

### How to reproduce the problem
Everything did work at one point. Things appeared to break after I rearranged my drives to use different ports. Not sure how that could result in this issue, but it does correlate.

### Warning/errors/backtraces from the system logs
An excerpt from dmesg
```
systemd[1]: zfs.target: Cannot add dependency job, ignoring: Unit zfs.target is masked.
```

zpool status claims the pool is online, reporting only that 'zpool upgrade' is available since upgrading to zfs 0.8.0.

zfs list correctly reports usage of the pool and subpools.

It's as though the mountpoint itself is the only failure.

**Comment by behlendorf** [comment]
> It's as though the mountpoint itself is the only failure.

This could be the case as boot time importing the pool and mounting the datasets are done as two steps.  I'd suggest starting by checking the status of the `zfs-mount.service` and `zfs-iimport.service`.  After a boot, if the pool is imported but not mounted you should also be able to run `zfs mount -a` to mount everything.  This is what the `zfs-mount.service` does, so running it manually may provide a clue as to why it's failing.

**Comment by bazald** [comment]
Things seem to be a little strange, @behlendorf . It appears as though zfs.target, zfs-import.target, and zfs-mount.service were all masked. Not sure if that's at all normal?

1. I initially unmasked only zfs-mount.service using systemctl. This did not resolve the issue, but both `systemctl start zfs-mount.service` and `zfs mount -a` appeared to be capable of activating the mountpoint.

2. I've since unmasked both zfs.target and zfs-import.target as well, but all it's served to do is to break the ability of `systemctl start zfs-mount.service` to activate the mountpoint. `zfs mount -a` still appears to work, but I'm no closer to getting it to function at boot, before (or after) the services that depend on it.

**Comment by bazald** [comment]
This problem persists. If anyone might have new ideas, they would be most welcome.

**Comment by bazald** [comment, marked as answer]
This being moved to Q&A prodded me into Googling again. Upgrading to the latest release through Debian backports and following the instructions at https://askubuntu.com/questions/1186283/how-can-i-make-zfs-datasets-mount-automatically-on-boot seems to have resolved the issue for me at last.

Copied verbatim, `systemctl enable zfs.target zfs-import.service zfs-mount.service` and since `zfs.target` seems to be invalid, `systemctl enable zfs-import.service zfs-mount.service` should be equivalent.
